kernel/comps/keyboard/src/i8042_chip/controller.rs

```diff
@@ -0,0 +1,277 @@
+
+    // Set the controller configuration byte.
+    let mut config = controller.read_configuration()?;
+    config.remove(
+        Configuration::FIRST_PORT_INTERRUPT_ENABLED
+            | Configuration::FIRST_PORT_TRANSLATION_ENABLED
+            | Configuration::SECOND_PORT_INTERRUPT_ENABLED,
+    );
+
+    // Perform controller self-test.
+    controller.wait_and_send_command(Command::TestController)?;
+    let result = controller.wait_and_recv_data()?;
+    if result != 0x55 {
+        // Any value other than 0x55 indicates a self-test fail.
+        return Err(I8042ControllerError::ControllerTestFailed);
+    }
+    // The self-test may reset the controller. Restore the original configuration.
+    controller.write_configuration(&config)?;
+
+    // Determine if there are two channels.
+    controller.wait_and_send_command(Command::EnableSecondPort)?;
+    let has_second_port = config.contains(Configuration::SECOND_PORT_CLOCK_DISABLED)
+        && !controller
+            .read_configuration()?
+            .contains(Configuration::SECOND_PORT_CLOCK_DISABLED);
+    controller.wait_and_send_command(Command::DisableSecondPort)?;
+    // Flush the output buffer again since we may have enabled the second port.
+    while controller.receive_data().is_some() {}
+
+    // Perform interface tests to the first PS/2 port.
+    controller.wait_and_send_command(Command::TestFirstPort)?;
+    let result = controller.wait_and_recv_data()?;
+    if result != 0x00 {
+        return Err(I8042ControllerError::FirstPortTestFailed);
+    }
+
+    // Perform interface tests to the second PS/2 port (if it exists).
+    if has_second_port {
+        controller.wait_and_send_command(Command::TestSecondPort)?;
+        let result = controller.wait_and_recv_data()?;
+        if result != 0x00 {
+            return Err(I8042ControllerError::SecondPortTestFailed);
+        }
+    }
+
+    // Enable the first PS/2 port.
+    controller.wait_and_send_command(Command::EnableFirstPort)?;
+    if let Err(err) = super::keyboard::init(&mut controller) {
+        log::warn!("i8042 keyboard initialization failed: {:?}", err);
+        controller.wait_and_send_command(Command::DisableFirstPort)?;
+    } else {
+        config.remove(Configuration::FIRST_PORT_CLOCK_DISABLED);
+        config.insert(
+            Configuration::FIRST_PORT_INTERRUPT_ENABLED
+                | Configuration::FIRST_PORT_TRANSLATION_ENABLED,
+        );
+    }
+
+    // TODO: Add a mouse driver and enable the second PS/2 port (if it exists).
+
+    I8042_CONTROLLER.call_once(|| SpinLock::new(controller));
+    let mut controller = I8042_CONTROLLER.get().unwrap().lock();
+    // Write the new configuration to enable the interrupts after setting up `I8042_CONTROLLER`.
+    controller.write_configuration(&config)?;
+    // Flush the output buffer to ensure that new data can trigger interrupts.
+    while controller.receive_data().is_some() {}
+    Ok(())
```

Comment:
Just a nit, maybe make the final `Ok(())` a separate code block since it's not related to the stuffs in the last code block.

The same for the final `Ok(())` in `keyboard::init`.